[EduNotes] Explain possible causes and solutions for type inference timeouts (#32084)
apple/swift: userdocs/diagnostics/expression-too-complex.md

# Type Inference in Complex Expressions

> For some complex expressions, the Swift compiler is unable to infer types in reasonable time. This can occur for a number of reasons, but some common ones include:

Reviewer: Nit: We're really defining what it means for an expression to be "too complex" here, so perhaps best to avoid using the term in the first sentence of the definition. (Though, strictly, we're defining "_too_ complex" rather than "complex"; but the idea is the same.)

```suggestion
For some expressions, the Swift compiler is unable to infer types in reasonable time. This can occur for a number of reasons, but some common ones include:
```

- The expression makes use of many overloaded operators and/or functions.
- The expression contains literals which can be used as values of many different types (e.g. integer and floating point literals).

> - The expression includes a collection literal with a large number of nontrivial element expressions.

Reviewer: What does "nontrivial" mean here? We use "trivial" sometimes to designate certain POD types, but I'd imagine that's not what you're talking about in this situation.

Reviewer 2: I think `notrivial` could be removed here because it really depends on number of elements rather than on their complexity in `large collection literals`.

> These types of expressions can take a long time to typecheck because the compiler must consider and eliminate a larger number of possible solutions when assigning types to subexpressions.

Reviewer 3: Some examples would probably be nice, though they'd become out of date if we can fix the exponential type checking issues.

Author: I'll see if I can come up with a reasonable example or two. I'd definitely like to avoid anything that might become out of date to avoid confusion, since every release tends to bring improvements in this area.

Reviewer 2: We have a number of examples of flow type-checking in `validation-test/Sema/type_checker_perf/slow`.

Author: I added an example based on SR-12267 which is fairly realistic

When this error occurs, possible solutions include:

> - Breaking the expression up into multiple subexpressions by storing intermediate values as `let` constants. These simpler subexpressions can have their types inferred more quickly.

Reviewer: Nit: The juxtaposition of two prepositions ("up into") can be difficult to parse for non-native English speakers, particularly since "breaking up" is a more informal expression with several possible dictionary definitions. Consider instead "Breaking up the expression into..." or perhaps more simply, "Separating the expression into..."

> - Providing manual type annotations for parts of the expression. This reduces the number of types the compiler must infer, and as a result the number of possible solutions it must consider.

Reviewer: ```suggestion
- Providing manual type annotations for parts of the expression. This reduces the number of possible solutions that the compiler must consider.
```

It is self-evidently true that annotating with an explicit type means that the compiler no longer has to infer an implicit type.

Users are likely to be familiar with using a colon to indicate an explicit type on the LHS of an assignment, but I think in general many are not so familiar with using the `as` operator to indicate an explicit type for _parts_ of an expression, so it could be helpful to provide a simple example:

> For example, you can write `let array = [1 as Int, 2, 3]`.

Author: In the example I added, I showed ways to fix the problem with either `as` or an initializer call, which I think is sufficient

- Reducing the number of overloaded declarations. New overloads of common operators like `+` or global functions like `max` may lead to slower type inference in some contexts.

> Sometimes, implementation limitations may cause the compiler to be unable to infer types for simpler expressions as well. If this appears to be the case, please consider filing a bug at <bugs.swift.org> with an example that reproduces the issue.

Reviewer 3: It doesn't look like this link was properly linked

> Sometimes, implementation limitations may cause the compiler to be unable to infer types for simpler expressions as well. If this appears to be the case, please consider filing a bug at [bugs.swift.org](http://bugs.swift.org) with an example that reproduces the issue.

Reviewer: Perhaps some guidance as to what sorts of expressions qualify as "simpler"?

Reviewer 2: I consider overwhelming majority of cases of `too complex` errors to be compiler bugs. It would be good to re-phrase this to ask developers to submit bug reports for any instances of such behavior, that would allow us to provide a workaround and get test-case added to the suite so we can verify against it later.

Author: Sounds good to me!
